### DISCUSSION TITLE
Is there a different == null behavior for Span<T>?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/roslyn#48758] [Q&A] [answered]
Currently we are translating many of our functions to `Span<T>` and we found that there is a different behavior for Span`<T>` comparing against `null`.

Comparing regular structs to `null `always results in false, where in `Span<T>` it's somehow like using `IsEmpty`.

Is there a reason why this is done (enable a more transistion from arrays to `Span<T>`?). Obviously it's not possible to enable the same behavior for regular developers.

**Comment by alrz** [comment, marked as answer]
Span has an implicit conversion from `T[]` that's what null is passed to, therefore the behavior you are observing.

**Reply by msedi** [reply]
@alrz: Have you tested this? I was told that every struct comparison against null is automatically returning false (see here: https://github.com/dotnet/roslyn/issues/45030).

As you said there is an automatic conversion but does the compiler automatically infer a T[] from null?

I did some tests with this struct:

``` CS
    public ref struct S<T>
    {
        public static bool operator ==(S<T> left, S<T> right) => left == right;
        public static bool operator !=(S<T> left, S<T> right) => !(left == right);

        public static implicit operator T[](S<T> s) => new T[0];
    }
```

and it shows that the == operator is not even called, also the operator T[] is not called when doing this:

``` CS
         var s = new S<int>();
         Console.WriteLine(s == null);

```

**Reply by alrz** [reply]
You can see it in action here: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJ9KBhSgb1MpeqpwBZKBZACgGUADhAB2AHjQiYAPkoQAlI2asVaAGaVeESgF4dlEQFcMGRUxIqVAX2UsbJK0A==

Notice the call to `op_Implicit(!0[])`, that's the conversion from array.

**Reply by msedi** [reply]
Thanks. I think I got it.
